README.md
# hydroPSO
[![Research software impact](http://depsy.org/api/package/cran/hydroPSO/badge.svg)](http://depsy.org/package/r/hydroPSO) [![CRAN](http://www.r-pkg.org/badges/version/hydroPSO)](https://cran.r-project.org/package=hydroPSO) [![License](https://img.shields.io/badge/license-GPL%20%28%3E=%203%29-lightgrey.svg?style=flat)](http://www.gnu.org/licenses/gpl-3.0.html) [![monthly](http://cranlogs.r-pkg.org/badges/hydroPSO)](https://www.rpackages.io/package/hydroPSO) [![total](http://cranlogs.r-pkg.org/badges/grand-total/hydroPSO)](https://www.rpackages.io/package/hydroPSO) [![Build Status](https://travis-ci.org/hzambran/hydroPSO.svg?branch=master)](https://travis-ci.org/hzambran/hydroPSO) [![dependencies](https://tinyverse.netlify.com/badge/hydroPSO)](https://CRAN.R-project.org/package=hydroPSO)

hydroPSO is a global optimisation R package implementing a state-of-the-art version of the Particle Swarm Optimisation (PSO) algorithm (SPSO-2011 and SPSO-2007 capable), with a special focus on the calibration of environmental models. 

hydroPSO is parallel-capable, to alleviate the computational burden of complex models.

hydroPSO is model-independent, allowing the user to easily interface any model code with the calibration engine (PSO), and includes a series of controlling options and PSO variants to fine-tune the performance of the optimisation engine. An advanced sensitivity analysis function together with user-friendly plotting summaries facilitate the interpretation and assessment of the calibration results. 

Bugs / comments / questions / collaboration of any kind are very welcomed.


## Articles using `hydroPSO`
  
| Year | Journal |  Model(s) / Application | Article |
| ---- | ------- | ----------------------- | --------|
| 2013 | EMS     | SWAT-2005, MODFLOW-2005 | [A model-independent Particle Swarm Optimisation software for model calibration](http://dx.doi.org/10.1016/j.envsoft.2013.01.004) |
| 2013 | IEEE    | Benchmark functions     | [Standard Particle Swarm Optimisation 2011 at CEC-2013: A baseline for future PSO improvements](http://ieeexplore.ieee.org/xpl/articleDetails.jsp?arnumber=6557848) |
| 2013 | JoH     | LISFLOOD                | [Hydrological evaluation of satellite-based rainfall estimates over the Volta and Baro-Akobo Basin](http://www.sciencedirect.com/science/article/pii/S0022169413005295) |
| 2014 | JCH     | MODFLOW2005-MT3DMS      | [Particle Swarm Optimization for inverse modeling of solute transport in fractured gneiss aquifer](http://www.sciencedirect.com/science/article/pii/S0169772214000862) |
| 2014 | JRSE    | SWAT                    | [SWAT model parameter calibration and uncertainty analysis using the hydroPSO R package in Nzoia Basin, Kenya](http://41.204.187.24:8080/bitstream/handle/123456789/2286/SWAT%20model%20parameter%20calibration%20and%20uncertainty.pdf?sequence=1&isAllowed=y) |
| 2014 | GMD     | WALRUS                  | [The Wageningen Lowland Runoff Simulator (WALRUS): a lumped rainfall-runoff model for catchments with shallow groundwater](https://doi.org/10.5194/gmd-7-2313-2014) |
| 2014 | HESS    | WALRUS                  | [The Wageningen Lowland Runoff Simulator (WALRUS): application to the Hupsel Brook catchment and the Cabauw polder](https://doi.org/10.5194/hess-18-4007-2014) |
| 2014 | HP      | Travel time distributions | [Consequences of mixing assumptions for time‐variable travel time distributions](https://doi.org/10.1002/hyp.10372) |
| 2015 | HP      | HBV                     | [A coupled hydrology-biogeochemistry model to simulate dissolved organic carbon exports from a permafrost‐influenced catchment](https://doi.org/10.1002/hyp.10566) |
| 2015 | HESS    | LISFLOOD                | [Global warming increases the frequency of river floods in Europe](https://doi.org/10.5194/hess-19-2247-2015) |
| 2015 | HESS    | LISFLOOD                | [A pan-African medium-range ensemble flood forecast system](https://doi.org/10.5194/hess-19-3365-2015) |
| 2015 | EE      | MARS-based              | [Hybrid PSO-MARS-based model for forecasting a successful growth cycle of the Spirulina platensis from experimental data in open raceway ponds](https://doi.org/10.1016/j.ecoleng.2015.04.064) |
| 2015 | MJ      | Malaria transmission    | [Predicting the impact of border control on malaria transmission: a simulated focal screen and treat campaign](https://doi.org/10.1186/s12936-015-0776-2) |
| 2016 | SC      | Stock Market            | [Natural combination to trade in the stock market](https://doi.org/10.1007/s00500-015-1652-2) |
| 2016 | EMS     | SWAT-VSA                | [Coupling the short-term global forecast system weather data with a variable source area hydrologic model](https://doi.org/10.1016/j.envsoft.2016.09.008) |
| 2016 | JoH-RS  | LISFLOOD                | [Assessing the role of uncertain precipitation estimates on the robustness of hydrological model parameters under highly variable climate conditions](https://doi.org/10.1016/j.ejrh.2016.09.003) |
| 2016 | NHESS   | LISFLOOD                | [Modelling the socio-economic impact of river floods in Europe](https://doi.org/10.5194/nhess-16-1401-2016) |
| 2017 | EP      | WALRUS-paddy+PDP        | [Hydrology and phosphorus transport simulation in a lowland polder by a coupled modeling system](https://doi.org/10.1016/j.envpol.2016.09.093) |
| 2017 | HP      | SWAT                    | [The value of remotely sensed surface soil moisture for model calibration using SWAT](https://doi.org/10.1002/hyp.11219) |
| 2017 | IS:CLS  | Genetics                | [Reconstructing Genetic Regulatory Networks Using Two-Step Algorithms with the Differential Equation Models of Neural Networks](https://doi.org/10.1007/s12539-017-0254-3) |
| 2017 | Bioener.| EPIC                    | [The greenhouse gas intensity and potential biofuel production capacity of maize stover harvest in the US Midwest](https://doi.org/10.1111/gcbb.12473) |
| 2017 | Sustain.| SWAT, GSWAT | [Development of an Evapotranspiration Data Assimilation Technique for Streamflow Estimates: A Case Study in a Semi-Arid Region](https://doi.org/10.3390/su9101658) |
| 2017 | CSR     | Clustering colors       | [Clustering colors](https://doi.org/10.1016/j.cogsys.2017.05.004) |
| 2017 | PLoS ONE| Partitioning of color space | [Does optimal partitioning of color space account for universal color categorization?](https://doi.org/10.1371/journal.pone.0178083) |
| 2017 | HESS    | Isotope analysis        | [Pesticide fate on catchment scale: conceptual modelling of stream CSIA data](https://doi.org/10.5194/hess-21-5243-2017) |
| 2017 | HESS (in review) | Dissolved organic carbon | [Hydrological control of dissolved organic carbon dynamics in a rehabilitated Sphagnum-dominated peatland: a water-table based modelling approach](https://doi.org/10.5194/hess-2017-578) |
| 2018 | Antrop. | WALRUS                  | [Hydrologic impacts of changing land use and climate in the Veneto lowlands of Italy](https://doi.org/10.1016/j.ancene.2018.04.001) |
| 2018 | JoH     | Soil moisture model (in R) | [Can next-generation soil data products improve soil moisture modelling at the continental scale? An assessment using a new microclimate package for the R programming environment](https://doi.org/10.1016/j.jhydrol.2018.04.040) |
| 2018 | AWM     | SWAT                    | [Assessing the impact of the MRBI program in a data limited Arkansas watershed using the SWAT model](https://doi.org/10.1016/j.agwat.2018.02.012) |
| 2018 | EMA     | Air quality             | [Air Quality Modeling Using the PSO-SVM-Based Approach, MLP Neural Network, and M5 Model Tree in the Metropolitan Area of Oviedo (Northern Spain)](https://doi.org/10.1007/s10666-017-9578-y) |



## Installation
Installing the latest stable version from [CRAN](https://CRAN.R-project.org/package=hydroPSO):
```{r}
install.packages("hydroPSO")
```

Alternatively, you can also try the under-development version from [Github](https://github.com/hzambran/hydroPSO):
```{r}
if (!require(devtools)) install.packages("devtools")
library(devtools)
install_github("hzambran/hydroPSO")
```

## Reporting bugs, requesting new features

If you find an error in some function, or want to report a typo in the documentation, or to request a new feature (and wish it be implemented :) you can do it [here](https://github.com/hzambran/hydroPSO/issues)


## Citation 
```{r}
citation("hydroPSO")
```

To cite hydroPSO in publications use:

> Zambrano-Bigiarini, M. and Rojas, R. (2013). [A model-independent Particle Swarm Optimisation software for model calibration](https://doi.org/10.1016/j.envsoft.2013.01.004), Environmental Modelling & Software, 43, 5-25, doi:10.1016/j.envsoft.2013.01.004.

> Zambrano-Bigiarini, M. and Rojas, R. (2018). [hydroPSO: Particle Swarm Optimisation, with Focus on Environmental Models](https://cran.r-project.org/package=hydroPSO). R package version 0.4-1. URL https://cran.r-project.org/package=hydroPSO. DOI:10.5281/zenodo.1287350.


BibTeX entries for LaTeX users are

>  @Article{Zambrano-BigiariniRojas2013-hydroPSO_article,
>      title = {A model-independent Particle Swarm Optimisation software for model calibration},
>      journal = {Environmental Modelling \& Software},
>      author = {Zambrano-Bigiarini, M. and Rojas, R.},
>      volume = {43},
>      pages = {5-25},
>      year = {2013},
>      doi = {10.1016/j.envsoft.2013.01.004},
>      url = {https://doi.org/10.1016/j.envsoft.2013.01.004},
>    }


>  @Manual{Zambrano-BigiariniRojas-hydroPSO_pkg,
>      title = {hydroPSO: Particle Swarm Optimisation, with Focus on Environmental Models},
>      author = {Mauricio Zambrano-Bigiarini and Rodrigo Rojas},
>      year = {2018},
>      note = {R package version 0.4-0. doi:10.5281/zenodo.1287350},
>      url = {https://CRAN.R-project.org/package=hydroPSO},

## Vignettes 
1) [Here](https://doi.org/10.5281/zenodo.3712382) you can find a vignette showing how to use `hydroPSO` to calibrate parameters of the **GR4J** hydrological model, which belongs to the [airGR](https://cran.r-project.org/package=airGR) family of models.

2) [Here](https://doi.org/10.5281/zenodo.3701902) you can find a vignette showing how to use `hydroPSO` to calibrate parameters of [TUWmodel](https://cran.r-project.org/package=TUWmodel).

3) [Here](https://doi.org/10.5281/zenodo.3701890) you can find a vignette showing how to use `hydroPSO` to calibrate parameters of SWAT-2005 and MODFLOW-2005.A similar approach can be used to calibrate SWAT-2012 or other models that need to be run from the system console.

   * The file [MF2005.zip](http://www.rforge.net/hydroPSO/files/MF2005.zip), with all the necessary files to run the MODFLOW-2005 examples in the vignette, contains **3 Windows binary files**: `mf2005.exe`, `preproc.exe` and `zonbud_hydroPSO.exe`. These binary files are distributed in the hope that they will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  In no event shall the authors be liable for any CLAIM, DAMAGES or other LIABILITY, whether in an action of contract, tort or otherwise, arising from, out of or in connection with the software or the use or other dealings in the software. 

   * The file [SWAT2005.zip](http://www.rforge.net/hydroPSO/files/SWAT2005.zip) , with all the necessary files to run the SWAT-2005 examples in the vignette, contains **1 Windows binary file**: `swat2005.exe` and  **1 UNIX binary file**: `swat2005.out`. Those binary files are distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  In no event shall the authors be liable for any CLAIM, DAMAGES or other LIABILITY, whether in an action of contract, tort or otherwise, arising from, out of or in connection with the software or the use or other dealings in the software. 


## Related Material 

* *hydroPSO: A Versatile Particle Swarm Optimisation R Package for Calibration of Environmental Models* (**EGU-2012**) [abstract](http://meetingorganizer.copernicus.org/EGU2012/EGU2012-10950-2.pdf), [poster](http://www.slideshare.net/hzambran/egu2012-10950hydro-pso4web).

* *R: a statistical environment for hydrological analysis* (**EGU-2010**) [abstract](http://meetingorganizer.copernicus.org/EGU2010/EGU2010-13008.pdf), [poster](http://www.slideshare.net/hzambran/egu2010-ra-statisticalenvironmentfordoinghydrologicalanalysis-9095709).



## See Also 

* [hydroGOF: Goodness-of-fit functions for comparison of simulated and observed hydrological time series](https://github.com/hzambran/hydroGOF).

* [hydroTSM: Time Series Management, analysis and interpolation for hydrological modelling](https://github.com/hzambran/hydroTSM).


## Feedback 
[Give us your opinion !](http://www.surveyexpression.com/Survey.aspx?id=dc5b906f-2bc7-400c-b8ea-23a4420a724f).


NEWS
NEWS for hydroPSO
--------------------------
0.5-9	Under-development
        o Now it passes TRAVIS-CI without any warning ('tcl' and 'tk' were added to .travis.yml to avoid error related to the installation of 'vioplot' and 'sm')
        o Source code has been partially modified in order to fix wrong display of IF/FOR/WHILE sections due to use of tabs instead of spaces
        o 'hydroPSO'     : -) When 'fn='hydromodInR', now all the arguments defined in 'model.FUN.args' are passed to the execution of 'model.FUN'. Existing vignettes of TUWmodel (v0.93) y GR4J (v0.3), because of the way the 'TUWhydromod' function was consruted in those vignettes.
                           -) Parallel execution now works correctly under Windows OS. Previously, errors like "object 'XX' not found" appeared when hydroPSO was run in parallel under Windows OS. This problem did not ocurred in GNU/Linux systems. Thanks Rodrigo Marinao for providing a solution.
        o 'verification' : -) now it is allows parallel computations ! (with progress bar)
                           -) When 'fn='hydromodInR', now all the arguments defined in 'model.FUN.args' are passed to the execution of 'model.FUN'. Existing vignettes of TUWmodel (v0.93) y GR4J (v0.3), because of the way the 'TUWhydromod' function was consruted in those vignettes.
                           -) Parallel execution now works correctly under Windows OS. Previously, errors like "object 'XX' not found" appeared when hydroPSO was run in parallel under Windows OS. This problem did not ocurred in GNU/Linux systems. Thanks Rodrigo Marinao for providing a solution.
        o 'hydromod'     : -) new argument 'exe.args' to allow passing additionl argument ot the model executable file (e.g., Raven https://cran.r-project.org/package=Rraven) .
                           -) now it allows to change parameterrs in three different ways: 'replacement' (the old and default way in hydroPSO), 'additive' (where the old value is summed up with the value computed during the optimisation), and 'multiplicative' (where the old value is multiplied by the value computed during the optimisation). Changes were carried out in the 'ParameterValues2InputFiles.R' and 'ModifyInputFile.R' files.
        o NAMESPACE      : 'pblapply' and 'pboptions' are now imported from the 'pbapply' package 


        

0.5-1	28-Apr-2020
        o Fixed error "object 'GoF' not found". This error appeared in version 0.5-0 when "fn" is neither "hydromod" nor "hydromodInR".
        o Updated [GR4J](https://CRAN.R-project.org/package=airGR) [vignette v0.3](https://doi.org/10.5281/zenodo.3774533), now including uncertainty in flow duration curves. 
        o Updated TUWmodel](https://CRAN.R-project.org/package=TUWmodel) [vignette v0.9.3](https://doi.org/10.5281/zenodo.3772176), now including uncertainty in flow duration curves.

 
0.5-0	17-March-2020
        o Package tested against R 4.0.0 (unstable) (2020-02-27 r77862) -- "Unsuffered Consequences", following an imperative request made by CRAN.
        o Added full compatibility with (hydrological/environmental) models implemented as R functions (e.g., TUWmodel, GR4, etc) 
        o New vignette showing how to calibrate TUWmodel with hydroPSO.
        o New vignette showing how to calibrate TUWmodel with GR4J.
        o New dataset 'Trancura9414001' with daily data on P, Temp, PET, and Q from 1979 to 2016.
        o A reference to the 2012 vignette showing how to use hydroPSO to calibrate SWAT2005 and MODFLOW 2005 is now included in the package.
        o 'hydroPSO'      : -) the 'fn' argument now accepts the character 'hydromodInR' to indicate that hydroPSO will be used to optimise an (hydrological/environmental) model implemented as R function. 
                            -) When 'fn="hydromodInR"', the argument 'model.FUN' must provide and R function used to run the model and evaluate the goodness-of-fit corresponding to a given parameter set.
                              This function MUST fulfill the following requirements:
                              1) It first agument must be named 'param.values', which represents the numer values of the parameters to be optimised
                              2) One of its arguments must be named 'obs', and represents the observed values whcih will be compared against the simulated values ('sim')
                              3) It must return a list object with -at least- the following elements:
                                 3.1) "GoF": the goodness-of-fit of obtained when the hydrological model is run with 'param.values' as parameters
                                 3.2) "sim": the model output obtained when the hydrological model is run with 'param.values' as parameters. 'sim' MUST have the same object class (and dimensions) thn the observed vlues provided by the user in 'obs'.
                              4) It is a good practice, even if is not a reuirement, to identify all the arguments used by the evaluation of the model, in order to have them correctly identified in the 'hydroPSO_logfile.txt' output file.
                            -) testing if 'parallel' package is installed is now optmised, replacing 'if ( !is.na( match("parallel", installed.packages()[,"Package"] ) ) )' by '( length(find.package("parallel", quiet=TRUE)) == 0 )'
                            -) testing if 'lhs' package is installed is now optmised, replacing 'if ( !is.na( match("lhs", installed.packages()[,"Package"] ) ) )' by '( length(find.package("lhs", quiet=TRUE)) == 0 )'
                            -) broken link in documtnation related to swarm topolies was fixed by using Lynn et al. (2018)
                            -) if 'multicore' option is selected, it is changed automatically to 'parallel' (because the 'multicore' pacakge was merged with the base 'parallel' pacakge.
        o 'verification'  : -) now it is fully compatible with R-based models and allows parallelisation

R/read_out.R
# File read_out.R
# Part of the hydroPSO R package, https://github.com/hzambran/hydroPSO
#                                 http://cran.r-project.org/web/packages/hydroPSO
#                                 http://www.rforge.net/hydroPSO/
# Copyright 2010-2020 Mauricio Zambrano-Bigiarini & Rodrigo Rojas
# Distributed under GPL 2 or later

################################################################################
#                            'read_out'                                        # 
################################################################################
# Purpose: This function reads the values of the objective function / model for#
#          each particle and iteration                                         #
################################################################################
# Output : list with two elements:                                             #
#          -) model.values: numeric or matrix/data.frame with the values of the#
#                           objective function / model for each particle and   #
#                           iteration                                          #
#          -) model.gofs  : numeric vector with the goodness-of-fit value for  #
#                           each particle and iteration                        #
#          -) model.best  : numeric with the best model output                 #
#          -) model.obs   : numeric with the observed values used for hydroPSO,#
#                           or NA if they were not provided                    #
################################################################################
# Author : Mauricio Zambrano-Bigiarini & Rodrigo Rojas                         #  
# Started: 08-Nov-2011,                                                        #
# Updates: 27-Jan-2012 ; 02-Feb-2012 ; 15-Feb-2012 ; 15-Feb-2012 ; 22-Feb-2012 #  
#          23-Mar-2012 ; 06-Dec-2012                                           # 
#          29-May-2013                                                         #  
#          09-Abr-2014                                                         # 
#          30-Jul-2015                                                         # 
#          28-Feb-2020 ; 07-Mar-2020 ; 12-Mar-2020                             #
################################################################################
# Columns in 'of_out' are:
# Iter         : integer, with the iteration number for each row of the file
# Part         : integer, with the particle number for each row of the file
# GoF          : numeric, with the goodness of fit value for each particle and iteration
# Model_Output : numeric vector (one or more values for each row), with the model output for each particle and iteration

read_out <- function(file="Model_out.txt", 
                     modelout.cols=NULL, 
                     nsim=NULL,
                     obs, 
                     MinMax=NULL, 
                     beh.thr=NA, 
                     verbose=TRUE,
                     #### Plotting arguments ####
                     plot=TRUE,
                     ptype=c("corr", "ts", "ecdf", "quant2ecdf"), 
                     ftype="dm", # OPTIONAL, only used when 'ptype=="ts"'.See [hydroGOF]{ggof}
                     FUN=mean,   # OPTIONAL, only used when 'ptype=="ts"'.See [hydroGOF]{ggof}                     
                     #### OPTIONS for ptype="quant2ecdf' #####
                     weights=NULL,
                     byrow=TRUE, 
                     quantiles.desired= c(0.05, 0.5, 0.95),
                     quantiles.labels= c("Q5", "Q50", "Q95"),
                     main=NULL,
                     ylab="Probability",
                     col="blue",
                     leg.cex=1.2,
                     leg.pos="bottomright",
                     cex.axis=1.2, 
                     cex.main=1.2, 
                     cex.lab=1.2,                     
                     #### PNG options ### 
                     do.png=FALSE, png.width=1500, png.height=900, png.res=90,
                     png.fname="ModelOut_vs_Obs.png" ) {                         
                         
  
  ##############################################################################
  # 1)                            Checkings                                    #
  ##############################################################################
  # Checking that 'file' exists
  if ( !file.exists(file) )
     stop( "Invalid argument value: The file '", basename(file), "' doesn't exist !!")

  # Checking 'MinMax'
  if ( !is.null(MinMax) ) {
     if ( !(MinMax %in% c("min", "max")) )
        stop("Invalid argument: 'MinMax' must be in c('min', 'max')")
  } else { # MinMax was not provided: it is read from the output 'PSO_logfile.txt' file
           # It assumes that 'PSO_logfile.txt' is located in the same directory than the 'Model_out.txt' file
      PSOlog <- read.table(file="PSO_logfile.txt", skip=9, nrows=1)
      MinMax <- as.character(PSOlog[1, 3])
      if (verbose) message("[ 'MinMax' was read from the 'PSO_logfile.txt' file, and set to '", MinMax, "' ]")
    } # ELSE end

  # Checking 'beh.thr'
  if ( !is.na(beh.thr) ) {
     if ( is.null(MinMax) ) {
        warning("Missing argument: 'MinMax' has to be provided before using 'beh.thr' !!")  
     } # IF end
  } # IF end
  
  # Checking 'plot' and 'MinMax'
  valid.types <- c("corr", "ts") 
  if (plot &  (length(which(!is.na(match(ptype, valid.types )))) <= 0) ) {
    if (is.null(MinMax)) {
      warning("Missing argument: 'MinMax' has to be provided before plotting when 'ptype' is in c('corr', 'ts') => 'plot=FALSE'")
      plot <- FALSE
    } # IF end
  } # IF end

  ##############################################################################
  # 2)                            Reading                                      #
  ##############################################################################

  # Reading the file
  if (verbose) message( "                                                     ")  
  if (verbose) message( "[ Reading the file '", basename(file), "' ... ]" )  
  if (!missing(nsim)) {
    cnames <- paste("sim", 1:nsim, sep="") 
    cnames <- c("Iter", "Part", "GoF", cnames)      
    sim  <- data.table::fread(file=file, header=FALSE, skip=1, fill=TRUE,  col.names=cnames, data.table=FALSE)
  } else sim  <- data.table::fread(file=file, header=FALSE, skip=1, fill=TRUE, data.table=FALSE)

  # Amount of total model outputs / parameter sets 
  nouts <- nrow(sim)
  if (verbose) message( "[ Total number of parameter sets: ", nouts, " ]" )       
  
  # computing the number of columns in 'file'
  ncols <- ncol(sim)  
  
  # Getting the GoF for each particle and iteration
  gofs <- sim[, 3]

  # Getting the Model Outputs for each particle and iteration
  outputs <- sim[, 4:ncols]
  
  # Computing the number of values in each model output
  lnsim  <- NCOL(outputs)
  if (verbose) message( "[ Number of model outputs for each parameter set ('nsim'): ", lnsim, " ]" )    
  
  # giving more meaningful names to the model outputs
  if (lnsim > 1)
    colnames(outputs) <- paste("Sim", 1:lnsim, sep="")
  
  # If the user only wants some columns of the model output file
  if (!is.null(modelout.cols)) {
    outputs <- outputs[, modelout.cols]
    
    lnsim  <- NCOL(outputs)
    if (verbose) message( "[ Number of extracted model outputs for each parameter set: ", lnsim, " ]" ) 
  } # IF end   
  
  # Filtering out those parameter sets above/below a certain threshold
  if (!is.na(beh.thr)) {  
     # Checking 'beh.thr'
     mx <- max(gofs, na.rm=TRUE)
      if (beh.thr > mx)
       stop("Invalid argument: 'beh.thr' must be lower than ", mx ,"!!")
    
    # Computing the row index of the behavioural parameter sets
    ifelse(MinMax=="min", beh.row.index <- which(gofs <= beh.thr), 
                          beh.row.index <- which(gofs >= beh.thr) )
    
    # Removing non-behavioural parameter sets and gofs
    if ( is.matrix(outputs) | is.data.frame(outputs) ) {
      outputs <- outputs[beh.row.index, ]
    } else outputs <- outputs[beh.row.index]
    gofs <- gofs[beh.row.index]
   
    # Amount of behavioural parameter sets 
    nbeh <- nrow(outputs)
    if (verbose) message( "[ Number of behavioural model outputs           : ", nbeh, " ]" ) 
    
  } # IF end
  
  # Selecting best model output
  if ( !is.null(MinMax) ) {
    ifelse(MinMax=="min", best.row.index <- which.min(gofs), 
                          best.row.index <- which.max(gofs) )
    if ( is.matrix(outputs) | is.data.frame(outputs) ) {
      best <- as.numeric(outputs[best.row.index, ]) 
    } else best <- as.numeric(outputs[best.row.index])
    
    # If the user only wants some columns of the model output file
    if (!is.null(modelout.cols)) best <- best[modelout.cols]
    
  } else {
      best <- NA
      plot <- FALSE
      if (verbose) warning("[ Note: You didn't provide 'MinMax' => model.best=NA & some plots will be skipped !! ]")
    } # ELSE end

  ##############################################################################
  # 3)                        Getting Observed Values                          #
  ##############################################################################

  if (missing(obs)) {
    fname <- "Observations.txt"
    if (file.exists(fname)) {
       if  ( length(find.package("zoo", quiet=TRUE)) != 0 ) {
        obs   <- read.zoo(fname) # zoo::read.zoo
        dates <- time(obs)
        # If the observed data do not have dates, they are transformed into numeric
        if ( TRUE && !(class(dates) %in% c("Date", "POSIXct", "POSIXt")) ) obs <- coredata(obs)
      } else {
        obs <- read.table(fname, header=FALSE, skip=0)  
        # skippping the column with the index
        obs <- obs[,1]
      } # ELSE end 
      
      # If the user only wants some columns of the model output file
      if (!is.null(modelout.cols)) {
        obs   <- obs[modelout.cols]
        dates <- dates[modelout.cols] 
      } # IF end
  
    } else obs <- NA
  } # IF end

  # Checking length(obs)
  if ( !is.null(obs) & is.numeric(obs) ) {
    if ( length(obs) != lnsim ) 
      stop("Invalid argument: 'length(obs) != ncol(sims)' ", length(obs), "!=", lnsim, " !!")
  } # IF end
  
  ##############################################################################
  # 4)                            Output                                       #
  ##############################################################################  
  # creating the output
  out <- list(model.values=outputs, model.gofs=gofs, model.best=best, model.obs=obs)

  ##############################################################################
  # 5)                            Plotting                                     #
  ##############################################################################
  if ( plot & is.numeric(obs) ) {
  
       plot_out(sim=best, 
                obs=obs, 
                dates=dates, 
                ptype=ptype, 
                MinMax=MinMax, 
                ftype=ftype, # OPTIONAL, only used when 'ptype=="ts"'.See [hydroGOF]{ggof}
                FUN=FUN,     # OPTIONAL, only used when 'ptype=="ts"'.See [hydroGOF]{ggof}
                verbose=verbose,
                #### OPTIONS for ptype="quant2ecdf' #####
                weights=weights,
                byrow=byrow, 
                quantiles.desired= quantiles.desired,
                quantiles.labels= quantiles.labels,
                main=main,
                ylab=ylab,
                col=col,
                leg.cex=leg.cex,
                leg.pos=leg.pos,
                cex.axis=cex.axis, 
                cex.main=cex.main, 
                cex.lab=cex.lab,                     
                #### PNG options ### 
                do.png=do.png, 
                png.width=png.width, 
                png.height=png.height, 
                png.res=png.res,
                png.fname=png.fname)
  } # IF end
  
  return(out)
  
}  # 'read_out' END
